packages/core/src/lib/base.ts

```diff
@@ -0,0 +1,319 @@
+ *
+ * @example
+ * class NodeFrontend extends FrontendBase<NodeBackend, NodeOptions> {
+ *   public constructor(options: NodeOptions) {
+ *     super(NodeBackend, options);
+ *   }
+ *
+ *   // ...
+ * }
+ */
+export abstract class FrontendBase<B extends Backend, O extends Options>
+  implements Frontend<O> {
+  /**
+   * The backend used to physically interact in the enviornment.
+   * Usually, this will correspond to the frontend. When composing SDKs,
+   * however, the Backend from the root SDK will be used.
+   */
+  private readonly backend: B;
+
+  /** Options passed to the SDK. */
+  private readonly options: O;
+
+  /**
+   * The client DSN, if specified in options. Without this DSN, the SDK will be
+   * disabled.
+   */
+  private dsn?: DSN;
+
+  /** A promise that resolves during installation. */
+  private installation?: Promise<boolean>;
+
+  /** Lazy cache of breadcrumbs, initialized when first accessed. */
+  private breadcrumbs?: Breadcrumb[];
+
+  /** Lazy cache of context data, initialized when first accessed. */
+  private context?: Context;
+
+  /**
+   * Initializes this frontend instance.
+   *
+   * @param backendClass A constructor function to create the backend.
+   * @param options Options for the frontend.
+   */
+  protected constructor(backendClass: BackendClass<B, O>, options: O) {
+    this.backend = new backendClass(this);
+    this.options = options;
+
+    if (options.dsn) {
+      this.dsn = new DSN(options.dsn);
+    }
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public async install(): Promise<boolean> {
+    if (!this.isEnabled()) {
+      return false;
+    }
+
+    if (this.installation) {
+      return this.installation;
+    }
+
+    this.installation = this.getBackend().install();
+    return this.installation;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public abstract captureException(exception: any): Promise<void>;
+
+  /**
+   * @inheritDoc
+   */
+  public abstract captureMessage(message: string): Promise<void>;
+
+  /**
+   * @inheritDoc
+   */
+  public async captureEvent(event: SentryEvent): Promise<void> {
+    await this.sendEvent(event);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public async addBreadcrumb(breadcrumb: Breadcrumb): Promise<void> {
+    const {
+      shouldAddBreadcrumb,
+      beforeBreadcrumb,
+      afterBreadcrumb,
+      maxBreadcrumbs = MAX_BREADCRUMBS,
+    } = this.getOptions();
+
+    if (maxBreadcrumbs === 0) {
+      return;
+    }
+
+    const timestamp = new Date().getTime() / 1000;
+    const mergedBreadcrumb = { timestamp, ...breadcrumb };
+    if (shouldAddBreadcrumb && !shouldAddBreadcrumb(mergedBreadcrumb)) {
+      return;
+    }
+
+    const finalBreadcrumb = beforeBreadcrumb
+      ? beforeBreadcrumb(mergedBreadcrumb)
+      : mergedBreadcrumb;
+
+    const breadcrumbs = await this.getBreadcrumbs();
+    breadcrumbs.push(finalBreadcrumb);
+    breadcrumbs.splice(0, breadcrumbs.length - maxBreadcrumbs);
```

Comment:
I'd always avoid mutation methods. It's very error-prone.

Reply:
This can be changed to

```js
breadcrumbs = [...breadcrumbs, finalBreadcrumb].slice(-maxBreadcrumbs);
```

Reply:
Ah right, breadcrumbs is no longer `readonly` :)